instrumentation/cassandra-driver/src/main/java/brave/cassandra/driver/TracingSession.java

```diff
@@ -0,0 +1,167 @@
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import zipkin.Endpoint;
+
+import static brave.Span.Kind.CLIENT;
+import static zipkin.internal.Util.checkNotNull;
+
+public final class TracingSession extends AbstractSession {
+  public static Session create(Tracing tracing, Session delegate) {
+    return new TracingSession(CassandraClientTracing.create(tracing), delegate);
+  }
+
+  public static Session create(CassandraClientTracing cassandraTracing, Session delegate) {
+    return new TracingSession(cassandraTracing, delegate);
+  }
+
+  final Tracer tracer;
+  final CassandraClientSampler sampler;
+  final CassandraClientParser parser;
+  final String remoteServiceName;
+  final TraceContext.Injector<Map<String, ByteBuffer>> injector;
+  final ProtocolVersion version;
+  final Session delegate;
+
+  TracingSession(CassandraClientTracing cassandraTracing, Session target) {
+    checkNotNull(cassandraTracing, "cassandraTracing");
+    this.delegate = checkNotNull(target, "delegate");
+    tracer = cassandraTracing.tracing().tracer();
+    sampler = cassandraTracing.sampler();
+    parser = cassandraTracing.parser();
+    String remoteServiceName = cassandraTracing.remoteServiceName();
+    this.remoteServiceName = remoteServiceName != null
+        ? remoteServiceName
+        : target.getCluster().getClusterName();
+    injector = cassandraTracing.tracing().propagation().injector((carrier, key, v) -> {
+      int length = v.length(); // all values are ascii
+      byte[] buf = new byte[length];
+      for (int i = 0; i < length; i++) {
+        buf[i] = (byte) v.charAt(i);
+      }
+      carrier.put(key, ByteBuffer.wrap(buf));
+    });
+    version = delegate.getCluster().getConfiguration().getProtocolOptions().getProtocolVersion();
+  }
+
+  @Override public ResultSetFuture executeAsync(Statement statement) {
+    Span span = nextSpan(statement);
+    if (!span.isNoop()) parser.request(statement, span.kind(CLIENT));
+
+    // o.a.c.tracing.Tracing.newSession must use the same propagation format
+    if (version.compareTo(ProtocolVersion.V4) >= 0) {
+      statement.enableTracing();
+      Map<String, ByteBuffer> payload = new LinkedHashMap<>();
+      if (statement.getOutgoingPayload() != null) {
+        payload.putAll(statement.getOutgoingPayload());
+      }
+      injector.inject(span.context(), payload);
+      statement.setOutgoingPayload(payload);
+    }
+
+    span.start();
+    ResultSetFuture result;
+    try {
+      result = delegate.executeAsync(statement);
+    } catch (RuntimeException | Error e) {
+      if (span.isNoop()) throw e;
+      parser.error(e, span);
+      span.finish();
+      throw e;
+    }
+    Futures.addCallback(result, new FutureCallback<ResultSet>() {
```

Comment:
have we any idea on the impact adding all the callbacks will cause.
EG in a client that processes 10k+ requests per second this could cause trouble, and a suitable load-shedding mechanism be appropriate.